plugins/catalog-backend-module-gitlab/src/lib/client.ts

```diff
@@ -63,6 +65,40 @@ export class GitLabClient {
     return this.pagedRequest(`/projects`, options);
   }
 
+  async listUsers(options?: ListOptions): Promise<PagedResponse<GitLabUser>> {
+    return this.pagedRequest(`/users`, options);
+  }
+
+  async listGroups(options?: ListOptions): Promise<PagedResponse<GitLabGroup>> {
+    return this.pagedRequest(`/groups`, options);
+  }
```

Comment:
These methods fetch all visible users and groups in the entire instance! This has me thinking two things:
1. It might be worth stating this in the documentation; a naive user might expect a config like
    ```yaml
    catalog:
      providers:
        gitlab:
          yourProviderId:
            host: my.gitlab.org
            group: mygroup
            orgEnabled: true
    ```
    to mean that the group `mygroup` and all of its descendant subgroups will be ingested, as well as all members of any group in this hierarchy. I'm pretty sure this is how #8917 worked.
1. In my gitlab instance, I can view ~15k users and ~5k groups. I'm attempting (still running, with a 4h timeout) to test this PR against that instance but it is slow, memory-intensive and sensitive to errors. If I could restrict the scope of entities ingested to descendants of a single group, it would make testing easier and could enable more stable production setups. It would be OK to add this feature later, though.

Reply:
While thinking about this, I discovered GitLab's [GraphQL API](https://docs.gitlab.com/ee/api/graphql) (generally available in GitLab 12.1+), and I found a query that could drastically cut down on the number of API calls:
```graphql
fragment BackstageUser on User {
  username
  name
  avatarUrl
  publicEmail
  bot
}

fragment BackstageGroup on Group {
  id
  path
  name
  description
  parent {
    id
  }
  groupMembers(relations: DIRECT) {
    nodes {
      user {
        ...BackstageUser
      }
    }
  }
}

fragment OrgData on Group {
  ...BackstageGroup
  descendantGroups {
    nodes {
      ...BackstageGroup
    }
  }
}

{
  group(fullPath: "{path from /api/v4/groups?top_level_only=true call}") {
    ...OrgData
  }
}
```

Reply:
Should we refactor that to GraphQL then?

Reply:
I think it doesn't have to be a blocker -- a switch like that, while beneficial, could happen later and stay backwards compatible. @benjdlambert wdyt?

Reply:
To be honest, i'm gonna defer to you folks about what your collective decision is. Can see pros and cons both ways, but happy to merge this and iterate. 

I'm not totally over the GitLab space, is 12.1 pretty old? Can we make this change knowing that it's not going to impact many people? Or would this be better off behind a config flag that people can opt-in and out of?

Reply:
I'd be happy to leave it as is now and make another PR with the GraphQL change.

Reply:
GitLab 12.1 was released July 2019 -- it looks like GitLab only [officially supports](https://about.gitlab.com/support/statement-of-support/#version-support) the 3 most recent major versions, which as of now are 13.x, 14.x and 15.x. So 12.x probably qualifies as "old" lol